calculator/calculator.py

```diff
@@ -1,218 +1,112 @@
+def operators_info_by_precedence() -> list:
+    """Sort the value of each item in OPERATORS by its priority.
 
     Returns:
-        A float, the result of the mathematical operation.
-
-    Note:
-        Brackets need a string operand (_evaluate() takes a string) and the
-        others a float (factorial and negate).
+        List of Operator named tuples by priority, strong to weak.
     """
-    if len(operands) == 2:
-        return operators[symbol].operation(float(operands[0]),
-                                           float(operands[1]))
-
-    operand = operands[0]
-    operation_to_do = operators[symbol].operation
-
-    if operation_to_do is not _evaluate:
-        operand = float(operands[0])
-
-    return operation_to_do(operand)
+    return sorted(OPERATORS.values(), key=operator.attrgetter("precedence"),
+                  reverse=True)
 
 
-def parse_sub_expression(symbol: str, sub_expr: str) -> list:
-    """Analyze a sub expression to its operands, by its operator.
+def replace_in_expression(operator_regex: str, expression: str,
+                          new_expr: str) -> str:
+    """Change the old sub expression to a new one according to pattern.
 
     Args:
-        symbol: the operator of the sub expression.
-        sub_expr: the expression to analyze.
+        operator_regex: pattern to find a sub expression of the operator.
+        expression: the entire expression.
+        new_expr: the expression to put instead of the old one.
 
     Returns:
-        A list of both operands, left and right.
+        A new whole expression, changed according to pattern.
     """
-    parsed_expr = sub_expr.split(symbol)
-
-    if symbol == "(":
-        parsed_expr[1] = parsed_expr[1][:-1]  # removes ")"
-
-    return parsed_expr
-
-
-def has_operations(expression: str) -> bool:
-    """Return True if the expression has operations in it else, false."""
-    for symbol in get_operators_by_precedence():
-        if operators[symbol].regex.search(expression):
-            return True
-
-    return False
+    old_expression = re.search(operator_regex, expression).group()
+    return expression.replace(old_expression, new_expr)
 
 
 def _evaluate(expression: str) -> float:
     """Calculate the value of a mathematical expression."""
-    for operator_symbol in get_operators_by_precedence():
-        if not has_operations(expression):
-            return float(expression)
+    expression = str(re.sub(r"\s+", "", expression))
+    operators_by_precedence = operators_info_by_precedence()
 
-        sub_expression = get_sub_expression_by_symbol(operator_symbol,
-                                                      expression)
-        if not sub_expression:
-            continue
+    for operator_info in operators_by_precedence:
+        operator_regex = operator_info.regex
 
-        operands = parse_sub_expression(operator_symbol, sub_expression)
-        result = calculate_sub_expression(operator_symbol, operands)
+        try:
```

Comment:
well if you had to use try and except, it means you had a bad implementation.
using groups is smart, but you can make it look way better

Reply:
Changing it to `if ... else...` statement is better?

Reply:
If match is not None, is a known template for regex searching.

Reply:
@yairigal do you mean there's a regex to find this pattern or just a regular if statement?